File: strategy/filter/stoch.py
from dataclasses import dataclass

from core.models.parameter import CategoricalParameter, Parameter, RandomParameter
from core.models.stoch import StochType
from strategy.filter.base import BaseFilter, FilterType


@dataclass(frozen=True)
class StochFilter(BaseFilter):
    type: FilterType = FilterType.Stoch
    stoch_type: Parameter = CategoricalParameter(StochType)
    period: Parameter = RandomParameter(13.0, 16.0, 1.0)
    k_period: Parameter = RandomParameter(1.0, 5.0, 1.0)
    d_period: Parameter = RandomParameter(3.0, 5.0, 1.0)

File: ta_lib/strategies/filter/src/stoch.rs
use base::{Filter, OHLCVSeries};
use core::{Comparator, Series};
use shared::{stoch_indicator, StochType};

pub struct StochFilter {
    stoch_type: StochType,
    period: usize,
    k_period: usize,
    d_period: usize,
}

impl StochFilter {
    pub fn new(stoch_type: StochType, period: f32, k_period: f32, d_period: f32) -> Self {
        Self {
            stoch_type,
            period: period as usize,
            k_period: k_period as usize,
            d_period: d_period as usize,
        }
    }
}

impl Filter for StochFilter {
    fn lookback(&self) -> usize {
        let adjusted_lookback = std::cmp::max(self.period, self.k_period);
        std::cmp::max(adjusted_lookback, self.d_period)
    }

    fn confirm(&self, data: &OHLCVSeries) -> (Series<bool>, Series<bool>) {
        let (k, d) = stoch_indicator(
            &self.stoch_type,
            data,
            self.period,
            self.k_period,
            self.d_period,
        );

        (k.sgt(&d), k.slt(&d))
    }
}
